src/graphene.cpp

```diff
@@ -0,0 +1,1482 @@
+            int nSizeBlock = pblock->GetBlockSize();
+            int nSizeGrapheneBlock = ::GetSerializeSize(grapheneBlock, SER_NETWORK, PROTOCOL_VERSION);
+
+            // If graphene block is larger than a regular block then send a regular block instead
+            if (nSizeGrapheneBlock > nSizeBlock)
+            {
+                pfrom->PushMessage(NetMsgType::BLOCK, *pblock);
+                LOG(GRAPHENE, "Sent regular block instead - graphene block size: %d vs block size: %d => peer: %s\n",
+                    nSizeGrapheneBlock, nSizeBlock, pfrom->GetLogName());
+            }
+            else
+            {
+                graphenedata.UpdateOutBound(nSizeGrapheneBlock, nSizeBlock);
+                pfrom->PushMessage(NetMsgType::GRAPHENEBLOCK, grapheneBlock);
+                LOG(GRAPHENE, "Sent graphene block - size: %d vs block size: %d => peer: %s\n", nSizeGrapheneBlock,
+                    nSizeBlock, pfrom->GetLogName());
+
+                graphenedata.UpdateFilter(grapheneBlock.pGrapheneSet->GetFilterSerializationSize());
+                graphenedata.UpdateIblt(grapheneBlock.pGrapheneSet->GetIbltSerializationSize());
+                graphenedata.UpdateRank(grapheneBlock.pGrapheneSet->GetRankSerializationSize());
+                graphenedata.UpdateGrapheneBlock(nSizeGrapheneBlock);
+                graphenedata.UpdateAdditionalTx(grapheneBlock.GetAdditionalTxSerializationSize());
+            }
+        }
+        catch (const std::runtime_error &e)
+        {
+            pfrom->PushMessage(NetMsgType::BLOCK, *pblock);
+            LOG(GRAPHENE,
+                "Sent regular block instead - encountered error when creating graphene block for peer %s: %s\n",
+                pfrom->GetLogName(), e.what());
+        }
+    }
+    else
+    {
+        dosMan.Misbehaving(pfrom, 100);
+
+        return;
+    }
+
+    pfrom->blocksSent += 1;
+}
+
+bool IsGrapheneBlockValid(CNode *pfrom, const CBlockHeader &header)
+{
+    // check block header
+    CValidationState state;
+    if (!CheckBlockHeader(header, state, true))
+    {
+        return error("Received invalid header for graphene block %s from peer %s", header.GetHash().ToString(),
+            pfrom->GetLogName());
+    }
+    if (state.Invalid())
+    {
+        return error("Received invalid header for graphene block %s from peer %s", header.GetHash().ToString(),
+            pfrom->GetLogName());
+    }
+
+    return true;
+}
+
+bool HandleGrapheneBlockRequest(CDataStream &vRecv, CNode *pfrom, const CChainParams &chainparams)
+{
+    if (!pfrom->GrapheneCapable())
+    {
+        dosMan.Misbehaving(pfrom, 100);
+        return error("Graphene block message received from a non graphene block node, peer %s\n", pfrom->GetLogName());
+    }
+
+    // Check for Misbehaving and DOS
+    // If they make more than 20 requests in 10 minutes then disconnect them
+    {
+        LOCK(cs_vNodes);
```

Comment:
I think this lock is unnecessary.  pfrom should have a reference added by the caller so no chance that it will be deleted